core/lib/snt/core/services/base.rb

```diff
@@ -0,0 +1,54 @@
+class SNT::Core::Services::Base
+  delegate :add_error, :add_validation_error, :add_validation_errors, :add_active_record_errors, :merge_result, :publish_events, to: :result
+
+  # Provide class level call method as convenience over calling new, then call
+  def self.call(*args, &block)
+    new(*args, &block).call
+  end
+
+  # @params
+  # - options [Hash] attributes:
+  #   - no_transaction [boolean] Do not start a transaction
+  # @return [Services::Result]
+  def call(options = {})
+    if options[:no_transaction]
+      # Don't start a transaction if no_transaction is true
+      call_delegate
+    else
+      # Start a transaction around the service call. Use requires_new to ensure sub transactions can be rolled back without impacting parent
+      # transaction. Services can optionally be invalidated and its transaction rolled back, when a sub-service fails and its sub-transaction is
+      # rolled back.
+      ActiveRecord::Base.transaction(requires_new: true) do
+        call_delegate
+      end
+    end
+
+    publish_events if ActiveRecord::Base.connection.open_transactions.zero? && result.status
```

Comment:
I do not see any way of hooking the commit into rails.  I think we should set a convention that a parent service should not be called when already in a transaction.  Services should be used to manage transactional changes.